<a href="https://github.com/lorenzo-arcioni/programmazione-python-base/blob/main/Capitolo2_Numeri_e_Stringhe/2_Aritmetica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ➕ Aritmetica in Python

In questo notebook vediamo come Python gestisce le **operazioni matematiche** di base e i comportamenti dei numeri nei calcoli.

Python supporta i classici operatori aritmetici usati in algebra:


| Operatore | Descrizione          | Esempio     | Risultato     |
|-----------|----------------------|-------------|----------------|
| `+`       | Addizione            | `5 + 3`     | `8`            |
| `-`       | Sottrazione          | `10 - 4`    | `6`            |
| `*`       | Moltiplicazione      | `2 * 7`     | `14`           |
| `/`       | Divisione (float)    | `8 / 3`     | `2.666...`     |
| `//`      | Divisione intera     | `8 // 3`    | `2`            |
| `%`       | Modulo (resto)       | `8 % 3`     | `2`            |
| `**`      | Potenza              | `2 ** 3`    | `8`            |


🧠 Python tratta i numeri come **oggetti** e le operazioni matematiche sono in realtà **metodi** definiti sui tipi numerici (`int`, `float`, `complex`, ecc.).

Per effettuare un'operazione in Python, come ad esempio $5 + 3$, ci basta eseguire semplicemente:

In [71]:
5 + 3

8

**Nota Bene**: Nei Jupyter Notebook, l'ultima istruzione di una cella viene stampata automaticamente anche senza `print()`. Questo negli script Python normali non succede, infatti per poter stampare il risultato di un'operazione è necessario utilizzare la funzione `print()`. 

In [72]:
print(5 + 3)

8


Possiamo assegnare il risultato di un'operazione ad una variabile, come ad esempio `somma = 5 + 3`:

In [73]:
somma = 5 + 3
print("Somma:", somma)

Somma: 8


### 🔢 Precedenza degli operatori

Come in matematica, alcuni operatori hanno precedenza su altri:

1. `**`
2. `*`, `/`, `//`, `%`
3. `+`, `-`

Puoi usare le **parentesi `()`** per forzare un ordine specifico.

In [74]:
print(2 + 3 * 4)       # 14 → moltiplicazione prima
print((2 + 3) * 4)     # 20 → parentesi prima
print(3 * 5 % 2)       # 1  → moltiplicazione prima
print(3 * (5 % 2))     # 3  → parentesi prima

14
20
1
3


## 🔢 Comportamento dei Tipi Numerici

Python gestisce automaticamente il tipo di numero in base al contesto dell’operazione:

- Se operi con due `int`, ottieni un `int` (salvo divisione `/`)
- Se uno degli operandi è `float`, il risultato sarà `float`
- Le divisioni con `/` **restituiscono sempre un `float`**, anche se il risultato è intero
- La divisione intera `//` **tronca** il risultato verso il basso

In [75]:
# Somma intera
print(4 - 3)

# Moltiplicazione intero-float
print(2 * 3.0)

# Divisione
print(6 / 3) # float

# Divisione intera
print(7 // 2)

1
6.0
2.0
3


## ⚠️ Casi Particolari: Divisione intera `//` e Modulo `%`

Python applica regole precise e coerenti anche con:
- **numeri negativi**
- **valori con la virgola (float)**

### 🔄 Divisione Intera con Float (`//`)

Anche se `//` restituisce un risultato **intero**, il **tipo del risultato** dipende dagli operandi:
- Se entrambi sono `int` → ritorna `int`
- Se almeno uno è `float` → ritorna `float` (senza parte decimale)

In [76]:
print(7 // 2)     # 3 (int)
print(7.0 // 2)   # 3.0 (float)
print(7 // 2.0)   # 3.0

3
3.0
3.0


### 📉 Divisione Intera con Numeri Negativi

Python **tronca verso il basso** (verso l’inf), non verso lo zero:

In [77]:
print(7 // 3)     # 2
print(-7 // 3)    # -3 (non -2)

2
-3


> Questo comportamento è diverso da alcuni altri linguaggi come C o Java, che troncano verso zero.

### 🧮 Modulo con Numeri Negativi

Il risultato del modulo `%` ha sempre il **segno del divisore**, non del dividendo:

In [78]:
print(7 % 3)      # 1
print(-7 % 3)     # 2
print(7 % -3)     # -2
print(-7 % -3)    # -1

1
2
-2
-1


### 🔬 Modulo e Divisione Intera con Float

Entrambe le operazioni sono **consentite con numeri decimali**, ma possono generare risultati inaspettati se non si è consapevoli del funzionamento:

In [79]:
print(7.5 // 2.0)   # 3.0
print(7.5 % 2.0)    # 1.5

print(-7.5 // 2.0)  # -4.0
print(-7.5 % 2.0)   # 0.5

3.0
1.5
-4.0
0.5


📌 **Regola generale**:
> Per ogni coppia `a`, `b` vale sempre la relazione:  
>  `a == (a // b) * b + (a % b)`

In [80]:
a = -7.5
b = 2.0
print(a == (a // b) * b + (a % b))  # True

True


📎 Questo comportamento è utile in algoritmi dove serve controllare **ripetizioni cicliche**, **allineamenti**, **calendari**, ecc.

## 🔢 Arrotondamenti e Conversione a intero

Python offre diversi strumenti per **gestire i numeri decimali**, a seconda di cosa vuoi ottenere:

| Funzione        | Descrizione                                 | Comportamento                       |
|----------------|----------------------------------------------|-------------------------------------|
| `round(x)`      | Arrotonda al numero intero più vicino        | Valore tondo più vicino (con .5 arrotonato verso il pari) |
| `round(x, n)`   | Arrotonda a `n` cifre decimali               | Utile per controllare la precisione |
| `int(x)`        | Converte a intero **troncando** la parte decimale | Non arrotonda, taglia!             |
| `abs(x)`        | Restituisce il valore assoluto del numero x | Toglie il segno al numero |

⚠️ Attenzione:
- `int(3.9)` → `3` (non 4!)
- `round(2.5)` → `2`, perché segue l'**arrotondamento bancario** (verso il numero pari)

In [91]:
print("round(3.14159):", round(3.14159))             # 3
print("round(3.14159, 2):", round(3.14159, 2))       # 3.14
print("int(3.99):", int(3.99))                       # 3
print("round(2.5):", round(2.5))                     # 2 (verso il pari)
print("round(3.5):", round(3.5))                     # 4
print(abs(3))                                        # 3
print(abs(-3))                                       # 3

round(3.14159): 3
round(3.14159, 2): 3.14
int(3.99): 3
round(2.5): 2
round(3.5): 4
3
3


## ⚖️ Perché `round()` in Python arrotonda al numero pari?

Quando usi `round(x)`, Python applica una regola particolare chiamata:

### 🔹 "Round Half To Even"
Conosciuto anche come **arrotondamento bancario**, è il metodo predefinito usato da `round()` per **ridurre l’errore cumulativo** nei calcoli numerici.

### 🧠 Come funziona?
Se un numero è esattamente a metà tra due interi (es. `2.5`, `3.5`, `4.5`...), Python **arrotonda verso il numero pari più vicino**:

| Numero        | Risultato `round(x)` |
|---------------|----------------------|
| `2.5`         | `2` ← pari            |
| `3.5`         | `4` ← pari            |
| `4.5`         | `4` ← pari            |
| `5.5`         | `6` ← pari            |

📌 Ma se **non è a metà esatta**, il numero viene arrotondato normalmente al più vicino:

| Numero        | Risultato `round(x)` |
|---------------|----------------------|
| `2.4`         | `2`                  |
| `2.6`         | `3`                  |

### ✅ Perché si fa così?
L’arrotondamento al pari è usato per **evitare distorsioni sistematiche** nei dati:
- Se si arrotondasse sempre per eccesso (`.5` → su), si introdurrebbe **un piccolo bias positivo** in grandi somme o medie.
- Arrotondare al pari **bilancia verso l’alto e verso il basso** in modo equo.

Questo metodo è adottato anche in linguaggi come Java, R, C# e nei fogli di calcolo (es. Excel, con `=ROUND()`).

In [82]:
print(round(2.5))  # 2
print(round(3.5))  # 4
print(round(4.5))  # 4
print(round(5.5))  # 6

2
4
4
6


## 🧮 Funzioni matematiche utili in Python

Python fornisce alcune **funzioni matematiche di base** integrate (senza moduli esterni):

- `abs(x)` → valore assoluto
- `pow(x, y)` → elevamento a potenza (`x ** y`)
- `divmod(x, y)` → restituisce una **tupla** `(x // y, x % y)` (divisione intera + resto)

In [83]:
print("abs(-10):", abs(-10))           # 10
print("pow(2, 3):", pow(2, 3))         # 8
print("2 ** 3:", 2 ** 3)               # 8 (equivalente)
print("divmod(17, 5):", divmod(17, 5)) # (3, 2)

abs(-10): 10
pow(2, 3): 8
2 ** 3: 8
divmod(17, 5): (3, 2)


`divmod()` è molto utile in operazioni dove servono **entrambi** i risultati della divisione: quoziente e resto.

## 🧠 Numeri Complessi

Python supporta nativamente i **numeri complessi**, usati in matematica, ingegneria, fisica, ecc.

- Un numero complesso è del tipo `a + bj`, dove:
  - `a` è la parte reale
  - `b` è la parte immaginaria
- Si crea con la lettera `j` (non `i`!)

In [84]:
c1 = 3 + 4j
c2 = 1 - 2j

print("Somma:", c1 + c2)
print("Prodotto:", c1 * c2)
print("Parte reale di c1:", c1.real)
print("Parte immaginaria di c1:", c1.imag)
print("Modulo di c1:", abs(c1))  # √(a² + b²)

Somma: (4+2j)
Prodotto: (11-2j)
Parte reale di c1: 3.0
Parte immaginaria di c1: 4.0
Modulo di c1: 5.0


## ⚠️ Limiti della rappresentazione dei numeri in virgola mobile (`float`)

In Python (come in quasi tutti i linguaggi), i numeri decimali (`float`) sono rappresentati **in virgola mobile** secondo lo standard **IEEE 754 a 64 bit**.

### 🧠 Cosa significa?
- Solo una quantità **finita** di numeri può essere rappresentata in modo **esatto**
- Alcuni numeri, come `0.1`, **non si possono rappresentare esattamente**
- Si introducono **approssimazioni**, che possono causare **piccoli errori nei calcoli**


### 🔎 Esempi pratici:
#### 1. **Errore di rappresentazione**

In [85]:
# Errore di rappresentazione
print("0.1 + 0.2 =", 0.1 + 0.2)  # Non sarà 0.3 esatto

0.1 + 0.2 = 0.30000000000000004


📉 **Risultato**: `0.30000000000000004`

Anche se **matematicamente** `0.1 + 0.2 = 0.3`, in binario questi numeri hanno una rappresentazione **infinita**, che viene **troncata**. L’errore è minimo, ma c’è!

#### 2. **Overflow** (numero troppo grande)

In [86]:
# Precisione persa con numeri molto grandi
print("1e308 * 10 =", 1e308 * 10)  # Inf (overflow)

1e308 * 10 = inf


📉 **Risultato**: `inf` (infinito)

Il valore `1e308` è già vicino al massimo rappresentabile (~`1.8e308`). Moltiplicandolo per 10, si supera il limite → Python restituisce `inf`.

#### 3. **Underflow** (numero troppo piccolo)

In [87]:
# Sotto il minimo rappresentabile
print("1e-324 / 10 =", 1e-324 / 10)  # 0.0 (underflow)

1e-324 / 10 = 0.0


📉 **Risultato**: `0.0`

Il numero `1e-324` è **il più piccolo numero positivo** rappresentabile in `float`. Dividendo ancora, il risultato diventa **così vicino a zero** che non può più essere distinto da zero stesso → Python restituisce `0.0`.

## 🔁 Assegnazioni combinate

In Python puoi usare **operatori abbreviati** per modificare il valore di una variabile direttamente, invece di riscrivere l'intera espressione.

### 📌 Esempi comuni:

| Operatore | Equivalente     | Effetto                       |
|-----------|------------------|-------------------------------|
| `+=`      | `x = x + y`      | somma e assegna               |
| `-=`      | `x = x - y`      | sottrai e assegna             |
| `*=`      | `x = x * y`      | moltiplica e assegna          |
| `/=`      | `x = x / y`      | dividi (float) e assegna      |
| `//=`     | `x = x // y`     | dividi intera e assegna       |
| `%=`      | `x = x % y`      | assegna il resto della divisione |
| `**=`     | `x = x ** y`     | eleva a potenza e assegna     |

Questi operatori rendono il codice più **conciso e leggibile**, specialmente nei cicli o nei contatori.

In [88]:
x = 10
x += 5     # x = x + 5 → 15
print(x)

x *= 2     # x = x * 2 → 30
print(x)

x %= 7     # x = x % 7 → 2
print(x)

15
30
2


## ➕➖ Operatori unari in Python

Python supporta anche gli **operatori unari**:

- `+x`: restituisce il valore **positivo** di `x` (in pratica, non cambia nulla)
- `-x`: restituisce l'**opposto** di `x` (lo cambia di segno)

> ❗ Non esistono operatori `++` o `--` come in C/C++ o Java. Scrivere `x++` dà errore.

### 🧠 Doppi operatori (`--x`, `+-x`, ecc.)
Python valuta da **sinistra a destra**:
- `--x` significa `-(-x)` → diventa **positivo**
- `+-x` significa `+(-x)` → resta **negativo**

Sono legittimi e possono essere usati, ma raramente sono necessari.

In [89]:
x = 5
print(+x)     # 5
print(-x)     # -5

print(--x)    # 5 → -(-5)
print(+-x)    # -5 → +(-5)

# Errore: Python non ha ++
# x++  ❌ SyntaxError

5
-5
5
-5


## 🔢 Conversioni tra basi numeriche

In Python puoi convertire numeri interi in **diverse basi**: binaria, ottale, esadecimale.

### 🔄 Funzioni utili:
- `bin(x)` → restituisce la **rappresentazione binaria** (`0b...`)
- `oct(x)` → restituisce l'**ottale** (`0o...`)
- `hex(x)` → restituisce l'**esadecimale** (`0x...`)
- `int(stringa, base)` → converte da **qualsiasi base a decimale**

### 🧠 Le stringhe di ritorno includono un prefisso (`0b`, `0o`, `0x`) che indica la base.

In [90]:
# Da decimale ad altre basi
x = 42
print(bin(x))   # '0b101010'
print(oct(x))   # '0o52'
print(hex(x))   # '0x2a'

# Da stringa di base N a intero decimale
print(int("101010", 2))   # 42 (binario → decimale)
print(int("2a", 16))      # 42 (esadecimale → decimale)
print(int("52", 8))       # 42 (ottale → decimale)

0b101010
0o52
0x2a
42
42
42


## ✅ Conclusioni

In questo notebook abbiamo esplorato in modo approfondito come Python gestisce le **operazioni aritmetiche** e il comportamento dei numeri.

Abbiamo visto:

- Come usare gli **operatori matematici** (`+`, `-`, `*`, `/`, `//`, `%`, `**`)
- Le **assegnazioni combinate** (`+=`, `*=` ecc.)
- Gli **operatori unari** (`+x`, `-x`, `--x`, ecc.)
- La gestione dei **numeri complessi**
- L’arrotondamento e le funzioni utili come `round()`, `abs()`, `divmod()`, `pow()`
- Le **conversioni tra basi numeriche** (binaria, ottale, esadecimale)
- I limiti della rappresentazione numerica: **overflow, underflow, e precisione**

💡 Python tratta i numeri come **oggetti** e consente di eseguire calcoli in modo molto flessibile, ma è importante conoscere le **caratteristiche tecniche** per evitare errori nei conti o risultati inattesi.

➡️ **Prosegui con il prossimo notebook sulle stringhe** per imparare a manipolare testi e caratteri in Python:

<a href="https://github.com/lorenzo-arcioni/programmazione-python-base/tree/main/Capitolo2_Numeri_e_Stringhe/3_Stringhe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>